### PR TITLE
[YUNIKORN-548] Allow AppID reuse after the app is completed

### PR DESCRIPTION
**Pull request by kingamarton**
The completed apps are moved into a separate map

### CODE REVIEW

**Comment by kingamarton** [comment]
First I wanted to append the date and time to the app ID when it was completed, but it is causing a lot of troubles around locking, so I removed it. We have the submission time, so we can differentiate the apps with the same ID according to their submission time.

File: pkg/scheduler/objects/application.go [hunk 1]
```diff
@@ -71,6 +71,7 @@ type Application struct {
 
 	rmEventHandler handler.EventHandler
 	rmID           string
+	callback       func(appID string)
```
Comment (yangwwei) [inline pkg/scheduler/objects/application.go:74]: Need a more explicit callback name for this function, easier for people to understand what this is for

**Review by yangwwei — COMMENTED** [review]
Looks good. Just add one small comment, and also please add some UT coverage for this callback mechanism. Thanks

File: pkg/scheduler/partition.go [hunk 2]
```diff
@@ -78,6 +79,7 @@ func newPartitionContext(conf configs.PartitionConfig, rmID string, cc *ClusterC
 		stateMachine: objects.NewObjectState(),
 		stateTime:    time.Now(),
 		applications: make(map[string]*objects.Application),
+		completedApplications: make(map[string]*objects.Application),
```
Comment (wilfred-s) [inline pkg/scheduler/partition.go:82]: please run `gofmt` over the file

File: pkg/scheduler/partition.go [hunk 3]
```diff
@@ -967,6 +980,11 @@ func (pc *PartitionContext) GetAppsByState(state string) []*objects.Application
 			appList = append(appList, app)
 		}
 	}
+	for _, app := range pc.completedApplications {
+		if app.CurrentState() == state {
+			appList = append(appList, app)
```
Comment (wilfred-s) [inline pkg/scheduler/partition.go:985]: We should not range over both lists. If the state is _completed_ we range over the `completedApplications` map in all other cases we range over the `applications` map.

File: pkg/scheduler/partition.go [hunk 4]
```diff
@@ -1222,3 +1240,12 @@ func (pc *PartitionContext) GetNodeSortingPolicy() policies.SortingPolicy {
 	defer pc.RUnlock()
 	return pc.nodeSortingPolicy.PolicyType
 }
+
+func (pc *PartitionContext) moveCompletedApp(appID string) {
+	pc.Lock()
+	defer pc.Unlock()
+	if app, ok := pc.applications[appID]; ok {
+		delete(pc.applications, appID)
+		pc.completedApplications[appID] = app
```
Comment (wilfred-s) [inline pkg/scheduler/partition.go:1249]: We need to be careful: there is no reason to assume that the same application ID is not re-used multiple times in the time we keep the old application around. A simple time stamp addition will allow us to re-use the appID multiple times:
```
// new ID as completedApplications map key, use negative value to get a divider
newID := appID + strconv.FormatInt(-(time.Now()).Unix(), 10)
```
The `newID` is then only used as the key in the _completedApplications_ map. No updates in the application object.

**Review by wilfred-s — CHANGES_REQUESTED** [review]
we should rename the `callback` variable to something linked to the completed state like `completedCallback` to make it really clear what it does like Weiwei proposed

**Review by wilfred-s — APPROVED** [review]
Changes are looking good, will commit when the build returns a pass

**Pull request closed without merging**